Adds user setings: historyCount controls number of logs shown

telamonian opened the pull request:
This PR adds a schema file for user settings for this extension. `jupyterlab-git` now has a section in the Advanced Settings Editor that you can access via the Settings menu. I've also added a simple `historyCount` setting to get the git settings started. It controls how many logs are visible in the history tab.

telamonian commented:
This PR will require #423 in order to pass CI

telamonian commented on schema/plugin.json:
```diff
@@ -0,0 +1,15 @@
+      "type": "integer",
+      "title": "History count",
+      "description": "Number of (most recent) commits shown in the history log",
+      "default": 100
```
I'm fine with making this smaller, but why do you think that it should be as low as 10? Even with this setting there will still be no easy way for a user to fetch more commit logs (eg a "show more commits" button or something), so we should show a bunch by default.

I think that 50, or even 25, would be a more reasonable default than 10.

telamonian replied:
I've reduced this to 25 in the latest

fcollonval replied:
I tend to be conservative on behavior when I'm late maintainer 😉 Although personally I'm find with 25. Is it ok for you @jaipreet-s ?

jaipreet-s replied:
Yep, 25 sounds fine!

telamonian commented on src/components/GitWidget.tsx:
```diff
@@ -77,11 +88,25 @@ export class GitWidget extends Widget {
+
+    void registry.load(key).then(settings => {
+      this._settings = settings;
+      // this._settings.changed.connect(this._loadSettings, this);
```
It's maybe not so critical for the "example" setting in this PR, but in general when a user saves a change to their settings all associated widgets should update. I added this in a different branch of `jupyterlab-git` that I'm working on right now, but forgot to include it in this PR. I'll add it in now.

telamonian replied:
Done

fcollonval replied:
Thx

fcollonval commented:
Could you please correct the error?
```bash
[E 02:53:47.312 BrowserApp] Uncaught exception POST /foo/git/all_history?1571453627307 (127.0.0.1)
    HTTPServerRequest(protocol='http', host='127.0.0.1:8888', method='POST', uri='/foo/git/all_history?1571453627307', version='HTTP/1.1', remote_ip='127.0.0.1')
    Traceback (most recent call last):
      File "/home/travis/virtualenv/python3.6.7/lib/python3.6/site-packages/tornado/web.py", line 1697, in _execute
        result = method(*self.path_args, **self.path_kwargs)
      File "/home/travis/build/jupyterlab/jupyterlab-git/jupyterlab_git/handlers.py", line 56, in post
        history_count = body["history_count"]
    KeyError: 'history_count'
```

telamonian commented:
I "fixed" the error. The basic problem is that extensions with special assets like schemas or themes aren't getting loaded correctly by the labextension install in `setup.py`

telamonian commented:
Actually, it's something even weirder. For whatever reason, the version of `jupyterlab-git` labextension that gets installed as a result of the:

```
pip install --find-links=dist jupyterlab_git[test]
```

command is actually the `0.8.2` that's currently up on `npmjs`. No idea why

telamonian commented:
So it looks like the Travis CI is indeed picking up a cached version of a wheel of the Python package at [one point in the resolution of the `pip install ...` call](https://travis-ci.org/jupyterlab/jupyterlab-git/builds/600224471#L693):

```
...
Found link https://files.pythonhosted.org/packages/30/e6/6020026c4f42ea11e53c074512c45a17633c1ca66e60ac78332d9c9ae1f9/jupyterlab_git-0.8.1.tar.gz#sha256=cb5f677ea49a1e8838fda0f76439e76b70da02f57d30d32fb7bee87820a8b21f (from https://pypi.org/simple/jupyterlab-git/), version: 0.8.1

  Found link file:///home/travis/build/jupyterlab/jupyterlab-git/dist/jupyterlab_git-0.8.2.tar.gz, version: 0.8.2

  Local files found: /home/travis/build/jupyterlab/jupyterlab-git/dist/jupyterlab_git-0.8.2.tar.gz

  Using version 0.8.2 (newest of versions: 0.1.1, 0.1.2, 0.2.0, 0.2.2, 0.3.0, 0.4.4, 0.5.0, 0.6.0, 0.6.1, 0.8.0, 0.8.1, 0.8.2)

  Using cached wheel link: file:///home/travis/.cache/pip/wheels/c5/f6/de/dce6e75470e1bcc6e8e49dbea4bc41577ae4cfe1ac687bfb26/jupyterlab_git-0.8.2-cp36-none-any.whl

  Added jupyterlab_git[test] from file:///home/travis/.cache/pip/wheels/c5/f6/de/dce6e75470e1bcc6e8e49dbea4bc41577ae4cfe1ac687bfb26/jupyterlab_git-0.8.2-cp36-none-any.whl to build tracker '/tmp/pip-req-tracker-ptcif6yr'

  Removed jupyterlab_git[test] from file:///home/travis/.cache/pip/wheels/c5/f6/de/dce6e75470e1bcc6e8e49dbea4bc41577ae4cfe1ac687bfb26/jupyterlab_git-0.8.2-cp36-none-any.whl from build tracker '/tmp/pip-req-tracker-ptcif6yr'
...
```

Still not sure how this leads to the installation of the latest Python package and an old `labextension`, but removing the `cache` lines from `.travis.yml` does fix the problem.

telamonian commented:
Bit of a mess, but I got there in the end. The `schema` dir had to be added to `MANIFEST.in`, and I had to prevent Travis from caching the wheel for `jupyterlab_git` (which I did by changing the initial build command for the python package in the CI to `python setup.py bdist_wheel`). Those two things together appear to be the minimal changes needed to fix the CI failure.

@fcollonval I added a unittest test for `GitAllHistoryHandler` in `test-handlers.py`. Anything else this PR needs?

fcollonval commented:
Thx for digging into caching that can be messy.

Two comments:
- I switch to sdist to follow [Steven comment](https://github.com/jupyterlab/jupyterlab-git/pull/415#pullrequestreview-291104714) in order to ensure the conda package will build (for instance forgetting the schema file would have broken the conda package). So is it possible to go back to it instead of the wheel?
- It would be interesting to execute `jupyter labextension list` before building. So we could be notified of which version is to be included? And in parallel we should bump to Dev version after each release to avoid the troubles you encountered here. What do you think?

telamonian commented:
In the CI, I switched back to `sdist`, and I added a call to `jupyter labextension list`.

I think bumping versions post release is a good idea. It's been standard practice in core and a bunch of other extensions. In spite of my earlier speculation, I actually don't think it would have helped fix the trouble I ran into with an old version of labextension getting installed (on a low level, it had to do with the missing schema dir and how `labextension install` validates packages (and how it falls back when that validation fails)). The cache problem was then caused by the fact that `pip`'s wheel cache won't automatically rebuild if the only thing that changes is the labextension `.tgz`.

In light of the additions made in this PR and others (eg #426), I bumped the versions of both labextension and Python package to `0.9.0`.

Bundling the Typescript code into the Python package honestly smells like an anti-pattern to me. Beyond the issues I ran into, there's a bunch of other obvious problems. For example, every time there's a new release of the labextension, we'll also have to do a new release of the Python package, even if there's no changes to the Python code. A better solution would be to just call out to the `labextension` app in `setup.py` (`jupyter labextension install jupyterlab_git` for release, `jupyter labextension install .` for develop). But let's save hashing that out for another PR.

telamonian commented:
Whoops, I already opened a new issue about it. Though I think that the idea of calling `labextension` in `setup.py` probably does deserve its own issue
